### PR TITLE
Pass language to slugify and unslugify. (WIP)

### PR DESCRIPTION
**Pull request by felixfontein**
First try on implementing passing of language to `slugify` and `unslugify`; see discussion in #2273.

### CODE REVIEW

File: nikola/nikola.py [hunk 1]
```diff
@@ -1730,19 +1730,20 @@ def scan_posts(self, really=False, ignore_quit=False, quiet=False):
...
-                    else:
-                        slugged_tags.add(utils.slugify(tag))
                     self.posts_per_tag[tag].append(post)
                 for lang in self.config['TRANSLATIONS'].keys():
+                    for tag in post.tags_for_language(lang):
+                        _tag_slugified = utils.slugify(tag, lang)
+                        if _tag_slugified in slugged_tags[lang]:
+                            if tag not in self.posts_per_tag:
+                                # Tags that differ only in case
+                                other_tag = [existing for existing in self.posts_per_tag.keys() if utils.slugify(existing) == _tag_slugified][0]
```
Comment (Kwpolska) [inline nikola/nikola.py]: missing `, lang` here.

**Comment by ralsina** [comment]
Since slugify could be used by plugins, unless you want this to have to
wait until v8, you should give lang a default value.

**Comment by Kwpolska** [comment]
I’m :-1: for this. You just wasted your time implementing a tiny obscure feature.

**Comment by felixfontein** [comment]
@ralsina: I added a warning in that case. This warning will be useful while this is WIP.

@Kwpolska: I really want to have this feature for my blog, for me it's neither tiny nor obscure. And at least @davidak and @akrist seem to agree since they bothered to open an issue resp. write code for this.

**Comment by davidak** [comment]
It is important for a german site since it is the grammatically correct representation.

Wordpress supports it since Version 3.6:
http://www.perun.net/2013/04/17/wordpress-3-6-umlaute-im-titel-ohne-plugin-entschaerfen/

before, most people used a plugin.

To implement this as a plugin would be ok for me.

there is a python module that does what we want:
https://pypi.python.org/pypi/translitcodec/

```
>>> import translitcodec
>>> import codecs
>>> codecs.encode(u'Schöne Grüße aus Köln', 'translit/long')
'Schoene Gruesse aus Koeln'
```

**Comment by felixfontein** [comment]
@davidak: this Python module cannot be used for all languages since for most it would be simply wrong. Also, I wonder how well it works for languages such as Chinese and Japanese.

My goal is to not change Nikola's default behavior, but to just enable this to be changed by plugins without writing a really insane plugin (which does post preprocessing etc.). Such a plugin could use this Python module for the German language, though.

About WordPress: I wonder why they only support de_DE and not also de_CH and de_AT. Also, the current version of WordPress uses similar translations for Danish (da_DK).

**Comment by davidak** [comment]
>Such a plugin could use this Python module for the German language, though.

that was my intention.

**Comment by ralsina** [comment]
This looks ok to me, and changes no behaviour. If noone sees a bug in the next 24 hours, I'll merge.

**Comment by felixfontein** [comment]
@ralsina: maybe we should leave it open a bit longer, I'd like to do some more intense tests (like with a real blog :) ). I'm not sure yet when I'll have time for this, but probably not during the next 24h.

**Comment by Kwpolska** [comment]
(Note we currently use `unidecode` which is a module that does the same as `translitcodec`

**Comment by Kwpolska** [comment]
(Note we currently use `unidecode` which is a module that does the same as `translitcodec`, just with different opinions, and I’m pretty sure CJK text is complete garbage with both)

File: tests/test_utils.py [hunk 2]
```diff
@@ -26,6 +26,7 @@ def test_getting_metadata_from_content(self):
         post = dummy()
         post.source_path = 'file_with_metadata'
         post.metadata_path = 'file_with_metadata.meta'
+        post.default_lang = 'en'
```
Comment (Kwpolska) [inline tests/test_utils.py]: Would be nicer to make a special `post_dummy` object with `default_lang = 'en'` already.

**Comment by ralsina** [comment]
@felixfontain ok, let's give it a few days. I want to do a release before friday because I will then be unavailable for a few weeks.

File: tests/test_utils.py (hunk 2 above)
Comment (felixfontein) [inline tests/test_utils.py]: I'm using `dummy` now, since that's precisely used in these situations.

**Comment by felixfontein** [comment]
Seems to work fine. I created a mini plugin for testing, for those who are interested, which can be put into the `plugins/` folder of a blog:
The description file (`t.plugin`):
```
[Core]
Name = t
Module = t
```
The code (`t.py`):
```
# -*- coding: utf-8 -*-

from nikola.plugin_categories import SignalHandler
from nikola import utils
from unidecode import unidecode
import re


_slugify_strip_re = re.compile(r'[^+\w\s-]')
_slugify_hyphenate_re = re.compile(r'[-\s]+')


def my_slugify(value, lang=None, force=False):
    if not isinstance(value, utils.unicode_str):
        raise ValueError("Not a unicode object: {0}".format(value))
    value = value.replace('ä', 'ae').replace('ö', 'oe').replace('ü', 'ue').replace('ß', 'ss')
    value = value.replace('Ä', 'ae').replace('Ö', 'oe').replace('Ü', 'ue')
    value = utils.unicode_str(unidecode(value))
    value = _slugify_strip_re.sub('', value, re.UNICODE).strip().lower()
    return 'slugified-' + _slugify_hyphenate_re.sub('-', value, re.UNICODE)


class T(SignalHandler):  # Could also be any other plugin type which is initialized early enough.
    def set_site(self, site):
        super(T, self).set_site(site)
        utils.slugify = my_slugify
```
I recommend to remove the `'slugified-'` prefix for production use ;-)

So from my side, merging is fine.

**Comment by felixfontein** [comment]
@ralsina: do you want to merge, or should I just do it myself?

**Comment by Kwpolska** [comment]
Merged, let's hope it works.

**Comment by felixfontein** [comment]
Thanks for merging!

**Pull request merged**